### PR TITLE
Moving funds Bitcoin transaction

### PR DESCRIPTION
**Pull request by tomaszslabon**
#Refs: https://github.com/keep-network/keep-core/issues/3734.

This PR introduces the creation of Bitcoin moving funds transaction which allows funds to be transferred between source wallet and target wallets.

The creation of the Bitcoin moving funds transaction is the responsibility of `movingFundsAction` which is executed
after moving funds commitment has been submitted.
Executing the `movingFundsAction` consists of several steps:
- initial validation of the moving funds proposal against the Ethereum network
- waiting `32` Ethereum blocks to ensure the commitment transaction has permanently entered the blockchain and will not be removed by a chain reorganisation
- final validation of the moving funds proposal against the Ethereum network which is needed to ensure the commitment was not changed during waiting
- assembling the moving funds Bitcoin transaction
- signing the moving funds Bitcoin transaction
- broadcasting the moving funds Bitcoin transaction

The moving funds Bitcoin transaction consists of one input (main wallet UTXO of the source wallet) and one or more outputs (one output for each target wallet). The outputs are all P2WPKH.

The three moving funds transactions used in tests were actually broadcasted within the Bitcoin testnet network to get the final confirmation that the transactions were correctly contructed.

### CODE REVIEW

**Review by lukasz-zimnoch — COMMENTED** [review]
Looks good! Just some small comments and I'm happy to merge

File: pkg/tbtc/moving_funds.go [hunk 1]
```diff
@@ -70,25 +102,193 @@ func newMovingFundsAction(
...
+
+		return walletData.MovingFundsTargetWalletsCommitmentHash != [32]byte{}, nil
+	}
+
+	// Wait `32` blocks to make sure the transaction has not been reverted for
+	// some reason, e.g. due to a chain reorganization.
+	result, err := ethereum.WaitForBlockConfirmations(
+		blockCounter,
+		currentBlock,
+		32,
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply pkg/tbtc/moving_funds.go]: Done in 7bfc095e5ce8cdf5e2a8c721528429b6e280de46.

File: pkg/tbtc/moving_funds.go [hunk 2]
```diff
@@ -70,25 +102,193 @@ func newMovingFundsAction(
...
+
+	// Just in case. This should never happen.
+	if mfa.proposalExpiryBlock < mfa.signingTimeoutSafetyMarginBlocks {
+		return fmt.Errorf("invalid proposal expiry block")
+	}
+
+	movingFundsTx, err := mfa.transactionExecutor.signTransaction(
+		signTxLogger,
+		unsignedMovingFundsTx,
+		mfa.proposalProcessingStartBlock,
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply pkg/tbtc/moving_funds.go]: Done in 118ac0e703b4f6a50f02c06a23d4bb13aa2c82be.

File: pkg/tbtc/moving_funds.go [hunk 3]
```diff
@@ -136,3 +333,63 @@ func (mfa *movingFundsAction) wallet() wallet {
...
+	// The sum of all the outputs equal to the input value minus fee.
+	totalOutputValue := walletMainUtxo.Value - fee
+
+	// The amount of Bitcoin allocated to each output. Notice that the last
+	// output may have a slightly higher amount. It happens if the total output
+	// amount is not divisible by the number of target wallets. The amount for
+	// the last output will be increased by the remainder left from distributing
+	// the total output amount across all the outputs. The value of this
+	// remainder is at most `len(targetWallets) - 1` satoshis.
+	singleOutputValue := totalOutputValue / int64(len(targetWallets))
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply pkg/tbtc/moving_funds.go]: Done in 7850b84221cbbc516b72ee577674ec71c7b4560d.

File: pkg/tbtc/moving_funds.go [hunk 4]
```diff
@@ -136,3 +333,63 @@ func (mfa *movingFundsAction) wallet() wallet {
...
+
+	for i, targetWalletPublicKeyHash := range targetWallets {
```
[the remark this answers is not in the document]
Comment (tomaszslabon) [reply pkg/tbtc/moving_funds.go:386]: Done in c38e2c2ba34401791ca93f6d2e0a4410406bff79.
We don't have to check the order of target wallets as the created moving funds proposal contains them in the correct order.
The target wallets in the proposal are either [found](https://github.com/keep-network/keep-core/blob/c38e2c2ba34401791ca93f6d2e0a4410406bff79/pkg/tbtcpg/moving_funds.go#L215) or (retrieved)[https://github.com/keep-network/keep-core/blob/c38e2c2ba34401791ca93f6d2e0a4410406bff79/pkg/tbtcpg/moving_funds.go#L22].

When they are found, the order of target wallets is validated on commitment submission.
When they are retrieved, the order of target wallets is validated by checking if they have the same hash as commitment.

**Review by lukasz-zimnoch — APPROVED** [review]

**Pull request merged**